File: backend/TirthSeva.API/Controllers/BookingsController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;
using TirthSeva.API.DTOs;
using TirthSeva.API.Services;
using Microsoft.EntityFrameworkCore;
using TirthSeva.API.Models;

namespace TirthSeva.API.Controllers
{
    [Authorize]
    [Route("api/[controller]")]
    [ApiController]
    public class BookingsController : ControllerBase
    {
        private readonly BookingService _bookingService;
        private readonly IConfiguration _configuration;
        private readonly TirthSeva.API.Data.ApplicationDbContext _context;

        public BookingsController(BookingService bookingService, IConfiguration configuration, TirthSeva.API.Data.ApplicationDbContext context)
        {
            _bookingService = bookingService;
            _configuration = configuration;
            _context = context;
        }

        [HttpPost("validate")]
        public async Task<ActionResult> ValidateBooking([FromBody] CreateBookingRequest request)
        {
            var userIdClaim = User.FindFirst(ClaimTypes.NameIdentifier);
            if (userIdClaim == null)
            {
                return Unauthorized();
            }

            var isAvailable = await _bookingService.ValidateBookingAvailabilityAsync(request);
            
            if (!isAvailable)
            {
                return BadRequest(new { message = "Selected slot is no longer available" });
            }

            return Ok(new { message = "Slot is available" });
        }

        [HttpPost("confirm-and-book")]
        public async Task<ActionResult> ConfirmPaymentAndBook([FromBody] CreateBookingRequest request, [FromQuery] string transactionId)
        {
            var userIdClaim = User.FindFirst(ClaimTypes.NameIdentifier);
            if (userIdClaim == null)
            {
                return Unauthorized();
            }

            var userId = int.Parse(userIdClaim.Value);
            
            // Validate availability first (without decreasing occupancy)
            var isAvailable = await _bookingService.ValidateBookingAvailabilityAsync(request);
            if (!isAvailable)
            {
                return BadRequest(new { message = "Booking failed. Slot no longer available." });
            }

            // Create booking with payment completed status
            var booking = await _bookingService.CreateBookingWithPaymentAsync(request, userId);
            
            if (booking == null)
            {
                return BadRequest(new { message = "Booking failed. Slot no longer available." });
            }

            // Create payment record
            var payment = new TirthSeva.API.Models.Payment
            {
                BookingId = booking.Id,
                Amount = booking.TotalAmount,
                PaymentMethod = "Razorpay",
                TransactionId = transactionId,
                Status = "Success",
                CreatedAt = DateTime.UtcNow
            };

            _context.Payments.Add(payment);
            await _context.SaveChangesAsync();

            var bookingDto = await _bookingService.GetBookingByIdAsync(booking.Id);
            return Ok(bookingDto);
        }

        [HttpGet("{id}")]
        public async Task<ActionResult<BookingDetailDTO>> GetBookingById(int id)
        {
            var booking = await _bookingService.GetBookingByIdAsync(id);
            
            if (booking == null)
            {
                return NotFound();
            }

            return Ok(booking);
        }

        [HttpGet("my-bookings")]
        public async Task<ActionResult<List<BookingListDTO>>> GetMyBookings()
        {
            var userIdClaim = User.FindFirst(ClaimTypes.NameIdentifier);
            if (userIdClaim == null)
            {
                return Unauthorized();
            }

            var userId = int.Parse(userIdClaim.Value);
            var bookings = await _bookingService.GetUserBookingsAsync(userId);
            return Ok(bookings);
        }

        [Authorize(Roles = "ServiceProvider")]
        [HttpGet("provider-bookings")]
        public async Task<ActionResult<List<BookingListDTO>>> GetProviderBookings()
        {
            var userIdClaim = User.FindFirst(ClaimTypes.NameIdentifier);
            if (userIdClaim == null)
            {
                return Unauthorized();
            }

            var userId = int.Parse(userIdClaim.Value);
            var bookings = await _bookingService.GetProviderBookingsAsync(userId);
            return Ok(bookings);
        }

        [Authorize(Roles = "Admin")]
        [HttpGet("all")]
        public async Task<ActionResult<List<BookingDetailDTO>>> GetAllBookings()
        {
            var bookings = await _bookingService.GetAllBookingsAsync();
            return Ok(bookings);
        }

        [HttpDelete("{id}")]
        public async Task<ActionResult> CancelBooking(int id)
        {
            var userIdClaim = User.FindFirst(ClaimTypes.NameIdentifier);
            if (userIdClaim == null)
            {
                return Unauthorized();
            }

            var userId = int.Parse(userIdClaim.Value);
            var userRole = User.FindFirst(ClaimTypes.Role)?.Value;

            // Admin can cancel any booking, User can only cancel their own
            int? bookingUserId = userRole == "Admin" ? null : userId;

            var result = await _bookingService.CancelBookingAsync(id, bookingUserId);
            
            if (!result)
            {
                return NotFound();
            }

            return Ok(new { message = "Booking cancelled successfully" });
        }

        [Authorize(Roles = "ServiceProvider,Admin")]
        [HttpPut("{id}/status")]
        public async Task<ActionResult> UpdateBookingStatus(int id, [FromBody] UpdateBookingStatusRequest request)
        {
            var userIdClaim = User.FindFirst(ClaimTypes.NameIdentifier);
            var userRole = User.FindFirst(ClaimTypes.Role)?.Value;

            if (userIdClaim == null)
            {
                return Unauthorized();
            }

            var userId = int.Parse(userIdClaim.Value);
            
            // Admin can update any booking status, Provider can only update their own
            int? providerId = userRole == "Admin" ? null : userId;

            var result = await _bookingService.UpdateBookingStatusAsync(id, request.BookingStatus, providerId);
            
            if (!result)
            {
                return NotFound(new { message = "Booking not found or you don't have permission to update it" });
            }

            return Ok(new { message = $"Booking status updated to {request.BookingStatus}" });
        }


    }

    public class UpdateBookingStatusRequest
    {
        public string BookingStatus { get; set; } = string.Empty;
    }
}

File: backend-spring/src/main/java/com/tirthseva/api/controller/BookingsController.java
package com.tirthseva.api.controller;

import com.tirthseva.api.dto.booking.*;
import com.tirthseva.api.entity.Booking;
import com.tirthseva.api.entity.Payment;
import com.tirthseva.api.entity.User;
import com.tirthseva.api.repository.BookingRepository;
import com.tirthseva.api.repository.PaymentRepository;
import com.tirthseva.api.service.BookingService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.net.URI;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/bookings")
public class BookingsController {

    private final BookingService bookingService;
    private final BookingRepository bookingRepository;
    private final PaymentRepository paymentRepository;

    public BookingsController(BookingService bookingService, BookingRepository bookingRepository, PaymentRepository paymentRepository) {
        this.bookingService = bookingService;
        this.bookingRepository = bookingRepository;
        this.paymentRepository = paymentRepository;
    }

    @PostMapping("/validate")
    public ResponseEntity<?> validateBooking(@Valid @RequestBody CreateBookingRequest request,
            @AuthenticationPrincipal User user) {
        if (user == null) {
            return ResponseEntity.status(401).build();
        }

        boolean isAvailable = bookingService.validateBookingAvailability(request);
        
        if (!isAvailable) {
            return ResponseEntity.badRequest()
                    .body(Map.of("message", "Selected slot is no longer available"));
        }

        return ResponseEntity.ok(Map.of("message", "Slot is available"));
    }

    @PostMapping
    public ResponseEntity<?> createBooking(@Valid @RequestBody CreateBookingRequest request,
            @AuthenticationPrincipal User user) {
        if (user == null) {
            return ResponseEntity.status(401).build();
        }

        Booking booking = bookingService.createBooking(request, user.getId());

        if (booking == null) {
            return ResponseEntity.badRequest()
                    .body(Map.of("message", "Booking failed. Check availability."));
        }

        BookingDetailDTO bookingDto = bookingService.getBookingById(booking.getId());
        return ResponseEntity.created(URI.create("/api/bookings/" + booking.getId())).body(bookingDto);
    }

    @GetMapping("/{id}")
    public ResponseEntity<BookingDetailDTO> getBookingById(@PathVariable Integer id) {
        BookingDetailDTO booking = bookingService.getBookingById(id);

        if (booking == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(booking);
    }

    @GetMapping("/my-bookings")
    public ResponseEntity<List<BookingListDTO>> getMyBookings(@AuthenticationPrincipal User user) {
        if (user == null) {
            return ResponseEntity.status(401).build();
        }

        List<BookingListDTO> bookings = bookingService.getUserBookings(user.getId());
        return ResponseEntity.ok(bookings);
    }

    @PreAuthorize("hasRole('ServiceProvider')")
    @GetMapping("/provider-bookings")
    public ResponseEntity<List<BookingListDTO>> getProviderBookings(@AuthenticationPrincipal User user) {
        if (user == null) {
            return ResponseEntity.status(401).build();
        }

        List<BookingListDTO> bookings = bookingService.getProviderBookings(user.getId());
        return ResponseEntity.ok(bookings);
    }

    @PreAuthorize("hasRole('Admin')")
    @GetMapping("/all")
    public ResponseEntity<List<BookingDetailDTO>> getAllBookings() {
        List<BookingDetailDTO> bookings = bookingService.getAllBookings();
        return ResponseEntity.ok(bookings);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> cancelBooking(@PathVariable Integer id, @AuthenticationPrincipal User user) {
        if (user == null) {
            return ResponseEntity.status(401).build();
        }

        boolean isAdmin = user.getRole().equals("Admin");
        Integer bookingUserId = isAdmin ? null : user.getId();

        boolean result = bookingService.cancelBooking(id, bookingUserId);

        if (!result) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(Map.of("message", "Booking cancelled successfully"));
    }

    @PreAuthorize("hasAnyRole('ServiceProvider', 'Admin')")
    @PutMapping("/{id}/status")
    public ResponseEntity<?> updateBookingStatus(@PathVariable Integer id,
            @RequestBody UpdateBookingStatusRequest request,
            @AuthenticationPrincipal User user) {
        if (user == null) {
            return ResponseEntity.status(401).build();
        }

        boolean isAdmin = user.getRole().equals("Admin");
        Integer providerId = isAdmin ? null : user.getId();

        boolean result = bookingService.updateBookingStatus(id, request.getBookingStatus(), providerId);

        if (!result) {
            return ResponseEntity.status(404)
                    .body(Map.of("message", "Booking not found or you don't have permission to update it"));
        }

        return ResponseEntity.ok(Map.of("message", "Booking status updated to " + request.getBookingStatus()));
    }

    @PostMapping("/confirm-and-book")
    public ResponseEntity<?> confirmPaymentAndBook(@Valid @RequestBody CreateBookingRequest request,
            @RequestParam String transactionId,
            @AuthenticationPrincipal User user) {
        if (user == null) {
            return ResponseEntity.status(401).build();
        }

        // Create booking after successful payment
        Booking booking = bookingService.createBooking(request, user.getId());

        if (booking == null) {
            return ResponseEntity.badRequest()
                    .body(Map.of("message", "Booking failed. Slot no longer available."));
        }

        // Create payment record
        Payment payment = Payment.builder()
                .bookingId(booking.getId())
                .amount(booking.getTotalAmount())
                .paymentMethod("Razorpay")
                .transactionId(transactionId)
                .status("Success")
                .createdAt(LocalDateTime.now())
                .build();

        paymentRepository.save(payment);

        BookingDetailDTO bookingDto = bookingService.getBookingById(booking.getId());
        return ResponseEntity.ok(bookingDto);
    }
}
